Wikipedia/Code/ArticleCollectionViewCell.swift

```diff
@@ -148,4 +153,221 @@ open class ArticleCollectionViewCell: CollectionViewCell {
         }
     }
     
+    // MARK: - Swipeable
+    
+    var collectionView: UICollectionView? {
+        return self.superview as? UICollectionView
+    }
+    
+    public var actionsView: CollectionViewCellActionsView?
+    var privateContentView = UIView()
+    
+    var swipeType: CollectionViewCellSwipeType = .none
+    
+    var swipeInitialFramePosition: CGFloat = 0
+    var swipeStartPosition: CGPoint = .zero
+    var swipePastBounds: Bool = false
+    var deletePending: Bool = false
+    var swipeVelocity: CGFloat = 0
+    var originalStartPosition: CGPoint = .zero
+    
+    public weak var delegate: SwipeableDelegate?
+    
+    var swipeTranslation: CGFloat {
+        get {
+            let x = privateContentView.frame.origin.x
+            return x
+        }
+        set {
+            privateContentView.frame.origin.x = newValue
+            actionsView?.frame = actionsViewFrame
+        }
+    }
+    
+    var actionsViewFrame: CGRect {
+        let width = actionsView?.maximumWidth ?? 0
+        let x = privateContentView.frame.origin.x
+        return CGRect(x: privateContentView.frame.width + x, y: privateContentView.frame.origin.y, width: width, height: privateContentView.frame.height)
+    }
+    
+    var minimumSwipeTrackingPosition: CGFloat {
+        guard let actionsView = actionsView else { return 0 }
+        return -actionsView.maximumWidth
+    }
+    
+    public var actions: [CollectionViewCellAction] {
+        get {
+            return self.actionsView?.actions ?? []
+        }
+        set {
+            self.actionsView?.actions = newValue
+        }
+    }
+    
+    public var isSaved: Bool {
+        if let delegate = delegate, let indexPath = indexPathForActiveCell {
+            return delegate.isArticleSaved(at: indexPath)
```

Comment:
Rather than give the cell specific knowledge of save/unsaved state, the action controller could ask the delegate for which actions are available for a given row.